README.md
# SamaruX

## Overview

SamaruX is a Unix-like shell for the CP/M operating system & the Z80 CPU.

It has been developed with MESCC, my Small-C compiler version for CP/M and Z80.

Instead of write separate commands for each task as in real Unix, the shell has some built-in commands.

Of course, it's not at all a new idea (see [BusyBox](http://www.busybox.net) for example), but I believe this scheme has a lot of advantages in a small environment like CP/M.

For example, we can save some bytes on disk by joining some commands in one file, instead of writing them as separated ones, because we don't have to repeat some shared code in each of them.

Anyway, the last versions of SamaruX support the development of external commands too.

Obviously, SamaruX doesn't have all functionalities of Unix, but it has some nice and useful commands, to work in a Unix-like system, under our beloved CP/M operating system.

Some SamaruX functionalities are:
- Redirection of stdin and stdout with `>`, `<` and `>>`.
- Piping with |.
- Environment with variables `$TERM`, `$USER`, etc.
- Names for directories.
- Command line history.
- Profiling for environment customization.
- Command aliasing.
- Batch processing with flow control and arguments passing.
- Complete on-line manual.
- Prompt customization.
- Commands: `cp`, `rm`, `mv`, `man`, `cat`, `more`, `grep`, `cd` ...

This document is not an exhaustive reference of SamaruX, but only an overview, in order to show some of its capabilities.

Please, refer to the manual for a complete description.

## Builds

SamaruX comes as two flawors:
- `SX.COM` binary.
- `SX2.COM` binary.

### SX

This build configuration has almost every command as external. So it
supports external commands. Its main goal is to reduce the TPA
needed to run SamaruX. But it needs more disk space.

### SX2

This build configuration has every command as built-in. It does not
support external commands. Its main goal is to reduce the disk space
needed to run SamaruX. But it needs a bigger TPA.

## Named directories

SamaruX has support for named directories.

With the use of `diralias`, you can reference a directory (a drive and user specification) with a name. IE:

```
diralias system a0:
diralias temp m0:
diralias mescc a3:

cat mescc:sx.c | more
ls system:*.x
cd temp:
```

It's a good idea to include the diralias commands in the SamaruX profile.

## Executing SamaruX external commands

SamaruX will search the external command files in the directory specified in the environment variable `BINDIR`:

```
env BINDIR system:
```

It this variable does not exists, or it's not valid, SamaruX will search the external command files in the current working directory.

If you prefix the filename with a directory, Samarux will search the external command file there:

```
mescc:cc myprog
```

You can force SamaruX to search a external command file in the current working directory, if you prefix the filename with the character `:`, as in:

```
:cc
```

## Searching manuals with the command `man`

SamaruX will seach the manual files in the directory specified in the environment variable `MANPATH`:

```
env MANPATH manuals:
```

It this variable does not exists, or it's not valid, SamaruX will search the manual files in the current working directory.

If you prefix the filename with a directory, Samarux will search the manual file there:

```
man mescc:help stdio
```

You can force SamaruX to search a manual file in the current working directory, if you prefix the filename with the character `:`, as in:

```
man :help stdio
```

## Execution modes

There are two execution modes:
- CP/M mode: The SamaruX shell will execute only a command and will return to CP/M.
- Interactive mode: The SamaruX shell will read and execute your commands from the keyboard, until you return to CP/M.

CP/M mode example:

```
A>sx echo Hello world!
Hello world!

A>
```

Interactive mode example:

```
A>sx
Samarux v1.00 / 29 Dec 2014 - (c) FloppySoftware, Spain

CP/M version 2.2

28 built-in commands

Hi FloppySoftware welcome to Samarux!

[FloppySoftware at B00:] $ env
TERM = vt52
USER = FloppySoftware
PROMPT = [%u at %w] %$

[FloppySoftware at B00:] $ exit

A>
```

## Profiling

With profiling, you can customize your SamaruX environment.

Its use it's not compulsory, but highly recommended (and useful).

There are two files for profiling::

- profile.sx: For the interactive mode.
- profcpm.sx: For the CP/M mode.

An example of profiling in interactive mode:

```
# -------------------------- #
# SAMARUX START-UP PROFILE   #
# -------------------------- #
env TERM vt52
env USER FloppySoftware
env PROMPT [%u at %w] %$
#
alias h history 0
alias logout exit
#
echo Hi $USER welcome to Samarux!
echo
```

## Piping and redirection

With SamaruX, you can enter useful command lines like:

```
cat letter.txt | more
grep 'Gary Kildall' cpm.txt article.doc news.txt > refs.txt
ls -l *.h *.c > cfiles.txt
man cat | more
```

## Command line

You can separate commands in the same line with the semicolon `;`:

```
man cp > cp.txt ; clear ; cat cp.txt | more
```

You can include environment variables in your command lines:

```
env NAME Julius ; echo My name is $NAME
```

An argument can contain spaces if it is surrounded by single quotes `'`:

```
env NAME 'Julius Smith' ; echo My name is $NAME
```

SamaruX has a command line history facility.

To see the history, just type:

```
history
```

And the history command will reply something like:

```
[FloppySoftware at B00:] $ history
0: cat manual.doc | more
1: ed letter.txt
2: ls *.c
3: man cat
```

To select a history command entry, just type its number:

```
history 2
```

And the shell will present a command line, ready to be edited and / or executed:

```
[FloppySoftware at B00:] $ ls *.c
```

## Executing CP/M commands

You can execute CP/M commands by using the cpm command:

```
cpm PIP A:=M:*.COM
```

You will return to SamaruX, once the CP/M command had finished its work.

## Aliases

You can make aliases for your most frequently used commands:

```
alias dirall ls -f
```

Then, you can enter the defined alias as a single word:

```
dirall
```

## Small text editor

SamaruX has the `ed` command, a very humble text editor, but very useful to create and edit small text files in a fast way.

```
[FloppySoftware at B00:] $ ed profile.sx
File :  profile.sx
Lines:  14/48
ed> print
    0: # -------------------------- #
    1: # SAMARUX START-UP PROFILE   #
    2: # -------------------------- #
    3: # env HOME A00:
    4: env TERM vt52
    5: # env TEMP M00:
    6: env USER FloppySoftware
    7: env PROMPT [%u at %w] %$
    8: #
    9: alias h history 0
   10: alias logout exit
   11: #
   12: echo Hi $USER welcome to Samarux!
   13: echo
ed> edit 8
    8: # This is a comment
```

## Help!

SamaruX has a built-in command named man, to offer on-line help about commands an other topics.

The help contents is stored in the `samarux.man` file.

As this file is an ordinary text file with a simple structure, you can even add or modify topics.

To see the available topics, just type:

```
man | more
```

To read about a topic in particular, say `cat`, just type:

```
man cat | more
```

In addition to this, SamaruX has a built-in command named builtin, that prints the name of all available built-in commands. Just type:

```
builtin
```

## Batch processing and flow control

With the batch command, you can process files with commands as if they were typed on the keyboard.

Some other SamaruX commands will help you with flow control in batch processing:
- `if`
- `goto`
- `read`
- `exit`

An example:

```
# ----------- #
# TEST SCRIPT #
# ----------- #
#
echo Test Script
echo ===========
# Menu:
echo
echo 1 : Option One
echo 2 : Option Two
echo 0 : Exit
echo
echo -n Your choice:
read OP
echo
if $OP eq 1 goto One
if $OP eq 2 goto Two
if $OP eq 0 goto Exit
echo Bad choice, try again.
goto Menu
# One:
echo Your choice was Option One.
goto Menu
# Two:
echo Your choice was Option Two.
goto Menu
# Exit:
env OP
```

You can include arguments in batch processing.

If you have a file named `write.sx` with this contents:

```
echo $1 $2 $3
```

And you type:

```
batch write.sx How are you?
```

The system will reply with:

```
How are you?
```

## License

Copyright (c) Miguel I. Garcia Lopez / FloppySoftware, Spain

[www.floppysoftware.es](http://www.floppysoftware.es)

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.


history.txt
SamaruX

(c) Miguel I. Garcia Lopez / FloppySoftware

http://www.floppysoftware.es
https://cpm-connections.blogspot.com
floppysoftware@gmail.com


SamaruX v2.07, 21 Sep 2021
==========================

Last changes.

New built-in commands:
----------------------

expr : Compute simple expressions.
ps   : Print running commands.


SamaruX v2.06, 18 Dec 2016
==========================

Last changes.

Shell improvements:
-------------------

- Added support for '>>' in output redirection.

New built-in commands:
----------------------

fortune : Print a random adage.
sleep   : Suspend execution for a # of seconds.


SamaruX v2.05, 06 Jun 2016
==========================

Last changes.

New built-in commands:
----------------------

false  : Return false.
head   : Return first lines from files.
shift  : Shift positional parameters to the left.
tail   : Print last lines from files.
true   : Return true.
tty    : Return user's terminal name.
whoami : Print the user name.

External commands
-----------------

Some built-in commands can be compiled as external ones.

Compile SamaruX with SX_MINIMAL for a minimal number of built-in commands. Then
execute make_all_ext2 to build the excluded built-in commands as external.


SamaruX v2.04, 04 Sep 2015
==========================

Last changes.

Executing SamaruX external commands
-----------------------------------

Like in previous releases, SamaruX will search the external command files in the
directory specified in the environment variable BINDIR:

	env BINDIR system:

It this variable does not exists, or it's not valid, SamaruX will search the
external command files in the current working directory.

If you prefix the filename with a directory, Samarux will search the external
command file there:

	mescc:cc myprog

New in this release: You can force SamaruX to search a external command file in
the current working directory, if you prefix the filename with the character ':',
as in:

	:cc

Searching manuals with the command man
--------------------------------------

Like in previous releases, SamaruX will seach the manual files in the directory
specified in the environment variable MANPATH:

	env MANPATH manuals:

It this variable does not exists, or it's not valid, SamaruX will search the
manual files in the current working directory.

If you prefix the filename with a directory, Samarux will search the manual
file there:

	man mescc:help stdio

New in this release: You can force SamaruX to search a manual file in the current
working directory, if you prefix the filename with the character ':', as in:

	man :help stdio


SamaruX v2.03, 03 Sep 2015
==========================

Last changes.

New built-in command:
---------------------

diralias : Support for named directories.

With the use of diralias, you can reference a directory (a
drive and user specification) with a name.

I.E.:

diralias root a0:
diralias temp m0:
diralias mescc a3:

cat mescc:sx.c | more
ls root:*.x
cd root:


SamaruX v2.02, 22 Ago 2015
==========================

Last changes.

New examples of SamaruX external commands:
------------------------------------------

banner  - print text banners (ported)
cal     - print calendars (ported)
head    - print first lines of files (ported)
whoami  - print current user name
strings - print strings from binary files
robots  - a game (already included in last SamaruX version)

Added support for SamaruX external commands:
--------------------------------------------

As an example, the game robots has been included.


New built-in commands:
----------------------

date  : Print and set the system date and time (only CP/M 3).
mem   : Show memory usage.
sort  : Sort text lines.
wc    : Count text lines, words and characters.
tee   : Copy standard input to standard output and to a file.


Improved commands:
-----------------

batch : Default file type for batch files - ".sx".
cat   : Added -k and -n options.
cd    : Added support for $HOME.
cpm   : Under CP/M 3, it uses the BDOS fn. CHAIN TO PROGRAM.
ed    : Added COPY and PASTE commands.
env   : Added -s option.
if    : Added in_env test for environment variables.
man   : Added support for $MANPATH and external manuals.
        Default file type for manual files - ".man".


New environment variables:
--------------------------

- HOME    : Default user directory.
- BINDIR  : Directory for SamaruX external commands.
- MANPATH : Path for manuals.


SamaruX v1.02, 27 Feb 2015
==========================

Last changes.


New control characters for command line edition:
------------------------------------------------

To delete the entire line:

     CTRL-U (0x15)
     CTRL-X (0x18)

To retype the line:

     CTRL-R (0x12)


New environment variables:
--------------------------

- LINES   : Number of console lines.
- COLUMNS : Number of console columns.
- TMPDIR  : Temporary directory (used by pipes).

The use of these environment variables is optional.

When a builtin command asks for a system variable,
the procedure is as follows:

- If the corresponding environment variable exists and
  has a legal value, use it.

- Under CP/M 3, use the value from the SCB.

- Otherwise, use the default values (for the console dimensions,
  the values #defined in the source code; for the temporary
  drive, the current drive).


samarux.h
/*	samarux.h

	Header for SamaruX external commands with MESCC (Mike's Enhanced Small C Compiler
	for Z80 & CP/M).

	SamaruX external commands are PRL files, with a header of 16 bytes.

	Copyright (c) 2015-2021 Miguel I. Garcia Lopez

	This program is free software; you can redistribute it and/or modify it
	under the terms of the GNU General Public License as published by the
	Free Software Foundation; either version 2, or (at your option) any
	later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

	Notes:

	This header must to be included first in the source code.

	Don't include other standard MESCC headers (conio.h, string.h, etc.).
	
	Don't forget to update the version number and the linked symbols (see below),
	and to compile again the external commands, if the SamaruX shell changes.
	
	To update the linked symbols, after Samarux shell compilation:
	
	prntosym sx.prn > sx.sym
	
	and copy / paste the new symbol addresses.
	
	To compile a SamaruX external command (ie: robots.c):
	
	cc robots
	ccopt robots
	zmac robots.zsm --rel
	link robots [op]
	rename robots.prl robots.x

	Revisions:

	31 Mar 2015 : 1st version.
	01 Apr 2015 : Changes in code header.
	09 Apr 2015 : Idem.
	20 Aug 2015 : Added version number to header. Updated linked symbols for SamaruX v2.01.
	22 Aug 2015 : Updated to SamaruX v2.02.
	23 Aug 2015 : Added C declarations of stdin, stdout, stderr.
	24 Aug 2015 : Updated to SamaruX changes.
	04 Sep 2015 : Updated to SamaruX v2.03.
	29 Nov 2015 : Modified FILENAME_MAX value. Added external references to SamaruX symbols.
	24 Jan 2016 : Updated to SamaruX v2.04.
	03 Jun 2016 : Updated to SamaruX v2.05.
	27 Nov 2016 : Updated to SamaruX v2.06.
	17 Dec 2016 : Added support for rand.h library.
	12 Sep 2021 : Updated to SamaruX v2.07.
*/

#ifndef SAMARUX_H

#define SAMARUX_H

/* HOOKS
   -----
*/

#define exit HookExit

/* SAMARUX CODE HEADER
   -------------------
   16 bytes
*/

#asm
	CSEG

	JP 0

	DEFB 'SamaruX', 0
	
	; Version 2.07
	
	DEFB 2
	DEFB 07

	JP main
#endasm

/* MESCC.H
   -------
*/

#define BYTE	unsigned char
#define WORD	unsigned int
#define BOOL	char
#define NULL	0
#define TRUE	1
#define FALSE	0

#define SIZEOF_CHAR 1 /* [unsigned] char */
#define SIZEOF_INT  2 /* [unsigned] int */
#define SIZEOF_PTR  2 /* pointer */

/* CLOCK.H
   -------
*/

#define CLOCK_LEN 5 /* CP/M clock array lenght in BYTES - ie: unsigned char clk[5] */
#define DATE_LEN  7 /* Date array lenght in WORDS - ie: int dt[7] */

/* CPM.H
   -----
*/

#define BF_GETDRV 25
#define BF_SETDRV 14
#define BF_USER 32
#define BF_DMA 26
#define BF_FIND1ST 17
#define BF_FINDNEXT 18
#define BF_OSVER 12
#define BF_CONST 11
#define BF_FSIZE 35
#define BF_OPEN 15
#define BF_DELETE 19
#define BF_CREATE 22
#define BF_READSEQ 20
#define BF_WRITESEQ 21
#define BF_RENAME 23
#define BF_CLOSE 16
#define BF_ATTRIB 30

#define UX_FCX_SIZ 37
#define UX_FCX_USR 0
#define UX_FCX_DRV 1
#define UX_FCX_RRC 34 /* Random Record Number */

/* FILEIO.H
   --------
*/

#define FILE unsigned char
#define EOF -1

#define FILENAME_MAX 22  /* dusrname:filename.typ + ZERO */

extern FILE *stdin, *stdout, *stderr;

/* RAND.H
   ------
*/

#define RAND_MAX 32767

/* STDBOOL.H
   ---------
*/

#define bool  int
#define true  1
#define false 0

/* SAMARUX LINKED SYMBOLS
   ----------------------
*/

#include "sx.h"

extern int sv_cpmver;
extern int sv_drive;
extern int sv_user;

extern char sv_tmp_fn[FILENAME_MAX];

extern int *sv_hist;

extern int *sv_env_name;
extern int *sv_env_value;

extern int *sv_alias_name;
extern int *sv_alias_value;

extern int *sv_dir_path;
extern int *sv_dir_alias;

extern int sv_flg_hist;
extern int sv_flg_buf;
extern int sv_flg_quit;
extern int sv_flg_cpm;
extern int sv_batch;

extern BYTE sv_fcx[UX_FCX_SIZ];
extern BYTE sv_fcx2[UX_FCX_SIZ];

extern BYTE *sv_dma;

extern BYTE *sv_fcxbuf;

extern int sv_cmd_max;

//extern WORD sv_cmd_name[SX_MAX_CMD];
//extern WORD sv_cmd_fun[SX_MAX_CMD];

extern WORD *sv_run_name;

extern int sv_run_exit;

//extern char sv_ed_buf[SX_LINESIZE];

//extern char sv_pipe_in[FILENAME_MAX];
//extern char sv_pipe_out[FILENAME_MAX];

#include "sx_sym.h"

#endif
